tools/autograd/gen_python_functions.py

```diff
@@ -73,10 +73,9 @@
 if (r.isNone(${out_idx})) {
   ${call_dispatch}
 } else {
-  if (!r.isNone(${type_idx})) {
-    check_out_type_matches(r.tensor(${out_idx}), r.scalartype(${type_idx}), r.layout(${layout_idx}),
-                           r.device(${device_idx}), r.isNone(${device_idx}));
-  }
+  check_out_type_matches(r.tensor(${out_idx}), r.scalartype(${type_idx}), r.isNone(${type_idx}),
+                         r.layout(${layout_idx}), r.isNone(${layout_idx}),
+                         r.device(${device_idx}), r.isNone(${device_idx}));
```

Comment:
I guess, not for this PR, but this check should happen in ATen, shouldn't it?

Reply:
Usually in this case there's not a dtype in the _out method in ATen; when there is, it's checked there.